crates/transpiler/src/passes/alap_schedule_analysis.rs

```diff
@@ -0,0 +1,175 @@
+
+        let op_duration = *node_durations.get(&node_index.index()).ok_or_else(|| {
+            TranspilerError::new_err(format!(
+                "No duration for node index {} in node_durations",
+                node_index.index()
+            ))
+        })?;
+
+        let op_view = op.op.view();
+        let is_gate_or_delay = matches!(
+            op_view,
+            OperationRef::Gate(_)
+                | OperationRef::StandardGate(_)
+                | OperationRef::StandardInstruction(StandardInstruction::Delay(_))
+        );
+
+        // compute t0, t1: instruction interval, note that
+        // t0: start time of instruction
+        // t1: end time of instruction
+
+        let t1 = if is_gate_or_delay {
+            // Gate or Delay operation
+            let t0 = qargs
+                .iter()
+                .map(|q| *idle_before.get(q).unwrap_or(&0.0))
+                .max_by(|a, b| a.partial_cmp(b).unwrap())
+                .unwrap_or(0.0);
+            t0 + op_duration
+        } else if matches!(
+            op_view,
+            OperationRef::StandardInstruction(StandardInstruction::Measure)
+        ) {
+            // Measure operation
+            let t0 = qargs
+                .iter()
+                .chain(cargs.iter())
+                .map(|bit| *idle_before.get(bit).unwrap_or(&0.0))
+                .max_by(|a, b| a.partial_cmp(b).unwrap())
+                .unwrap_or(0.0);
+
+            //           |t1 = t0 + duration
+            //    Q ░░░░░▒▒▒▒▒▒▒▒▒▒▒
+            //    C ░░░░░░░░░▒▒▒▒▒▒▒
+            //               |t0 + (duration - clbit_write_latency)
+
+            let t1 = t0 + op_duration;
+            for clbit in cargs.iter() {
+                idle_before.insert(*clbit, t0 + (op_duration - clbit_write_latency as f64));
+            }
+            t1
+        } else {
+            // Directives (like Barrier)
+            let t0 = qargs
+                .iter()
+                .chain(cargs.iter())
+                .map(|bit| *idle_before.get(bit).unwrap_or(&0.0))
+                .max_by(|a, b| a.partial_cmp(b).unwrap())
+                .unwrap_or(0.0);
+            t0 + op_duration
+        };
+
+        for qubit in qargs {
+            idle_before.insert(qubit, t1);
+        }
+
+        node_start_time.insert(node_index, t1);
+    }
+
+    // Compute maximum instruction available time
+    let circuit_duration = *idle_before
+        .values()
+        .max_by(|a, b| a.partial_cmp(b).unwrap())
+        .unwrap_or(&0.0);
+    // Note that ALAP pass is inversely scheduled, thus
+    // t0 is computed by subtracting t1 from the entire circuit duration.
+    let py_dict = PyDict::new(py);
+    for (node_idx, t1) in node_start_time {
+        let node = dag.get_node(py, node_idx)?;
+        let time = circuit_duration - t1;
+        if time.fract() == 0.0 {
+            py_dict.set_item(node, time as u64)?;
+        } else {
+            py_dict.set_item(node, time)?;
+        }
+    }
```

Comment:
This step shouldn't be in a Rust native function. Though it is still necessary to achieve what you want. There should be a way of not making this conversion to a PyDict required.